### ISSUE TITLE
ACE2Pro connecting via ch340-based usb-rs485 cable (not stock ch343).

### ISSUE THREAD

**Issue by ArtDesignEngineering** [decay71/multiACE#46]
Hello.
I've got my ACE2Pro from Kobra X combo (flashed to 1.1.31 fw version already), so I do not have an Anycubic cable with usb-rs485 (ch343-based) converter.
I'm trying to connect my ACE2Pro using ordinary usb-rs485 (ch340-based) converter, I can see an ACE2Pro in Snapmaker's cli, it's name is:
_/dev/serial/by-path/platform-fed00000.usb-usb-0:1.3.3:1.0-port0_
And I can verify connectivity via my PC (I can see that ACE2Pro is connected and can update it's status in test .py script).
But I can't get it working with U1_1.4.1-paxx12-19_multiACE0.97.3b firmware, always getting **"Error [multiACE] USB unstable, expected 2 ACEs, found 1 - ACE inactive, Klipper continues. FIRMWARE_RESTART required after reconnecting"**
My ch340-based adapter has PID&VID=1a86:7523, and looks like MultiACE is hard-coded to PID&VID=%Anycubic ch343-based cable%. Can it be changed to use any ch340-based adapters?

**Comment by decay71** [comment]
Hi,

yes, good idea, will implement a list in the config section. For now here's a patched ace_protocol_v2.py for you that should to both.
Please confirm.

To get it working permanently you have to touch/oem/.debug and copy the file to /home/lava/klipper/klippy/extras/ace_protocol_v2.py if you used the bin flash.

[ace_protocol_v2.py](https://github.com/user-attachments/files/28941215/ace_protocol_v2.py)

**Comment by ArtDesignEngineering** [comment]
Unfortunately not working, got error "USB unstable, expected 2 ACEs, found 1 - ACE inactive, Klipper continues. FIRMWARE_RESTART required after reconnecting."
I have connected both ACE1 and ACE2 via USB hub and I can see them as:
```
root@lava:/dev/serial/by-path# lsusb
........
Bus 003 Device 006: ID 1a86:7523 QinHeng Electronics CH340 serial converter


root@lava:~# ls /dev/serial/by-id
usb-1a86_USB_Serial-if00-port0                                                   
usb-ANYCUBIC_ACE_1-if00         
                                                 
root@lava:~# ls /dev/serial/by-path
platform-fed00000.usb-usb-0:1.3.2.3:1.0      
platform-fed00000.usb-usb-0:1.3.4:1.0-port0 
```
And I have changed access rules for CH340 usb port:
`root@lava:~# chmod 666 platform-fed00000.usb-usb-0:1.3.4:1.0-port0`

multiace_usb.log shows me only ACE1-related events:
`2026-06-15 07:05:37.729 SCAN [startup] found=1 time=9.3ms devices=[/dev/serial/by-path/platform-fed00000.usb-usb-0:1.3.2.3:1.0(v1)->ttyACM1]`

klippy.log - too:
`07:05:57.960:[multiACE] Found device /dev/serial/by-path/platform-fed00000.usb-usb-0:1.3.2.3:1.0 (ttyACM1) protocol=v1
`

But when I'm connecting my ACE2Pro unit to my PC - I can check it's connectivity and status via _ace2_pro_shell.py_ tool script (and it means that my ch340 diy cable is working and it' pinout is correct):
```
ace2> info

→ GET_INFO seq=1  [ffaa00010007003b61fe]
  [10:32:49.041] 27 bytes
    0000: ff aa 80 01 00 07 11 0a 07 56 31 2e 31 2e 33 31   .........V1.1.31
    0010: 12 06 56 31 2e 30 2e 33 59 39 fe                  ..V1.0.3Y9.
  ← GET_INFO_RESP seq=1 flags=0x80
  version: V1.1.31
  boot_version: V1.0.3
ace2> discover

→ DISCOVER_DEVICE seq=2  [ffaa0002000000fe09fe]
  [10:33:05.798] 26 bytes
    0000: ff aa 80 02 00 00 10 08 9a a9 cb 91 03 10 b8 e0   ................
    0010: 10 18 d3 f2 c4 89 03 20 b1 fe                     ....... ..
  ← DISCOVER_DEVICE_RESP seq=2 flags=0x80
  uid: 0x3232D49A-0x00043038-0x31313953
  Device: 0x3232D49A-0x00043038-0x31313953
ace2> status

→ GET_STATUS seq=3  [ffaa00030006009541fe]
  [10:33:32.488] 28 bytes
    0000: ff aa 80 03 00 06 12 08 01 18 1a 20 26 28 01 30   ........... &(.0
    0010: 01 4a 00 4a 00 4a 00 4a 00 75 e6 fe               .J.J.J.J.u..
  ← GET_STATUS_RESP seq=3 flags=0x80
  work_state: IDLE
  temp: 26°C
  humidity: 38%
  rfid1: True
  rfid2: True
  slot[0]: READY / EMPTY
  slot[1]: READY / EMPTY
  slot[2]: READY / EMPTY
  slot[3]: READY / EMPTY
ace2>
```

**Comment by decay71** [comment]
Hi,

Two things were blocking it. First, the CH340's USB ID (1a86:7523) wasn't in the accepted list. But the real blocker is that the genuine ACE cable is a CDC-ACM device while your CH340 is a usb-serial (ttyUSB) adapter and multiACE read the USB vendor/product ID from a fixed sysfs path that only works for CDC-ACM. For a ttyUSB adapter that ID sits one directory level higher, so the read failed and your CH340 was silently skipped before it was ever checked. That's why just allowing 7523 wasn't enough. 

I guess so at least :)

Please try the attached file, as i don't have your adapter i can't test it myself.

[ace_protocol_v2.py](https://github.com/user-attachments/files/28947936/ace_protocol_v2.py)

**Comment by ArtDesignEngineering** [comment]
It works now, thank you!
Both ACE2Pro and ACE Pro are connected and operational right now.
How do you think, is it possible to stay on ch340 dongle, or it's better to move to ch343-based (for reliability/stability)?

**Comment by decay71** [comment]
Perfect! Will be implemented in next release.

I think  there's no fundamental CH340 vs CH343 difference, CH343 only pulls ahead at non-standard or very high baud rates, which don't apply here. 

So: staying on CH340 is OK in principle. I'd just validate it rather than trust it blindly - run a long real print  and check klippy.log for V2 CRC mismatch lines and any [Errno 5] / reconnects. If it stays clean over a few hours, it should be good. multiACE recovers from the occasional drop, but the V2 can't freewheel, so sustained comms loss = under-extrusion - that's the thing to rule out.

**Comment by ArtDesignEngineering** [comment]
And some more info: in my case there was no sufficient access rights for this serial port, so I have to create `/etc/udev/rules.d/99-anycubic-ace.rules` and add `SUBSYSTEM=="tty", ATTRS{idVendor}=="1a86", ATTRS{idProduct}=="7523", MODE="0666", GROUP="dialout"` in it, and apply with `udevadm control --reload-rules && udevadm trigger`

**Comment by decay71** [comment]
All right, will note that in the readme!

**Comment by ArtDesignEngineering** [comment]
And one more question please: how stable is ACE2Pro connectivity for now and is there any statistics about it?
I have now connected only one ACE2Pro unit with discussed ch340 dongle.
I have two prints right now, they both the same ~ 2 hours (it is the same gcode), the first one is printed very well, and the second one is failed approx.  at it's ~30% (it looks like massive under-extrusion because of jammed filament - maybe the feed_assist have not switched correctly and the whole layer was jammed). Unfortunately my options state_debug, usb_debug and fa_debug were turned off, so I have no logs, but I have checked ACEG__Status: `[multiACE] USB stats (uptime 201.1min): errno5=0 (recovered=0, lost=0), cascades=0, connects=1, disconnects=0` so looks like no any rs485 protocol errors at all. Klippy.log also shows no any errors.
So the question is - how to debug such an errors properly and what can be a reason of that?

**Comment by decay71** [comment]
There is no statistics tracking for this, but for what it is worth, I have printed well over 100 hours with my two ACE units without running into problems like this.

In the past there were feed-assist issues reported by some users, but those did not show up on every gcode, they were tied to specific files. Since you printed the exact same file twice, with the first run fine and only the second one failing, that particular failure pattern can be ruled out here. If it were one of that bugs, it would behave the same way on both runs.

I have also completed several prints of up to 12 hours single and mixed Ace 2 on the same version you are running, without issues.

To get a clearer picture of what happened, it would help to turn the logging options on (fa_debug, state_debug, usb_debug). Not all error paths are captured in the USB stats, so a clean status line alone does not rule everything out. With logging enabled, the next occurrence will give me something concrete to look at.
